packages/core/src/controllers/relayer.ts

```diff
@@ -135,20 +148,50 @@ export class Relayer extends IRelayer {
 
   public async transportClose() {
     this.transportExplicitlyClosed = true;
-    await this.provider.disconnect();
+    if (this.connected) await this.provider.disconnect();
+    this.events.emit(RELAYER_EVENTS.transport_closed);
   }
 
   public async transportOpen(relayUrl?: string) {
     this.relayUrl = relayUrl || this.relayUrl;
     this.transportExplicitlyClosed = false;
-    await this.provider.connect();
-    // wait for the subscriber to finish resubscribing to its topics
-    await new Promise<void>((resolve) => {
-      this.subscriber.once(SUBSCRIBER_EVENTS.resubscribed, () => {
-        resolve();
-      });
-    });
+    try {
+      await Promise.race([
+        this.provider.connect(),
+        new Promise<void>((_res, reject) =>
+          // rejects pending promise if transport is closed before connection is established
+          // useful when .connect() gets stuck resolving
+          this.once(RELAYER_EVENTS.transport_closed, () => {
+            reject();
+          }),
+        ),
+      ]);
+
+      if (this.initialized) {
+        // wait for the subscriber to finish resubscribing to its topics
+        await new Promise<void>((resolve) => {
+          this.subscriber.once(SUBSCRIBER_EVENTS.resubscribed, () => {
```

Comment:
Is there a possibility of `SUBSCRIBER_EVENTS.resubscribed` firing before the listener is added here, i.e. causing the flow to hang?

Reply:
I haven't seen that happening but as a precaution, I can move the event before the `.connect()` is called to be sure